Dtype Analysis for Unary and Binary ops with Metatensors

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67786
* #67575
* #67500
* #66973
* #66948
* #66899
* __->__ #66898

Differential Revision: [D32175961](https://our.internmc.facebook.com/intern/diff/D32175961)

Gamrix committed:
Dtype Analysis for Unary and Binary ops with Metatensors

[ghstack-poisoned]

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Krovatkin reviewed (commented):
looks pretty good! I have a few qs to make sure I understand the code. You could ignore all the "nit:" comments/suggestions/questions.

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Gamrix commented on test/jit/test_dtype_analysis.py:
```diff
@@ -0,0 +1,172 @@
+        ]
+
+        input_dtypes = [
+            (float32, float32),  # Simple Case
```
Yeah, I consider it covered by the combination of (float, float), and (int32, int64). 
(again, this test case is already taking multiple seconds, I am really worried about expanding it further).

Gamrix commented on test/jit/test_dtype_analysis.py:
```diff
@@ -0,0 +1,172 @@
+            (int32, float32),  # type Promotion
+        ]
+
+        with set_default_dtype(float32):
```
I don't see what we gain by that.

Krovatkin replied:
fair enough

Gamrix commented on torch/csrc/jit/passes/dtype_analysis.cpp:
```diff
@@ -0,0 +1,303 @@
+
+    switch (n->kind()) {
+      case aten::append:
+        // auto elementType =
```
Will actually just return False, as we will just not process that for now.

Gamrix commented on torch/csrc/jit/passes/dtype_analysis.cpp:
```diff
@@ -0,0 +1,303 @@
+
+static bool canBeInferredWithMetaTensor(Node* n) {
+  // Not a guarantee that the metatensor will not error out
+  // Do not have a allowlist for now and let things error out in execution.
```
Also, the custom rules are basically also intended to be a blocklist.

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Krovatkin commented on test/jit/test_dtype_analysis.py:
```diff
@@ -0,0 +1,172 @@
+        input_dtypes = [
+            (float32, float32),  # Simple Case
+            (int32, int64),  # Size Promotion (compliated case for 0dim tensors)
+            (float32, int32),  # type Promotion
```
makes sense.

Krovatkin commented on torch/csrc/jit/passes/dtype_analysis.cpp:
```diff
@@ -0,0 +1,303 @@
+    // in a previous run of dtype analysis (eg in testing with cached graphs)
+
+    // Due to SSA, there shouldn't be a reason why this overwrite is not valid.
+    value->setType(tensor_type->withScalarType(scalarType));
```
looks good

Krovatkin commented on test/jit/test_dtype_analysis.py:
```diff
@@ -0,0 +1,173 @@
+
+        # We need to clear shape information because torch.jit.script
+        # will return a cached graph if the function is scripted twice.
+        torch._C._jit_pass_erase_shape_information(graph)
```
awesome!!!!

Krovatkin reviewed (approved):
:shipit:

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"




[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175961).

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"


Differential Revision: [D32175961](https://our.internmc.facebook.com/intern/diff/D32175961)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175961).

Gamrix committed:
Update on "Dtype Analysis for Unary and Binary ops with Metatensors"


Differential Revision: [D32175961](https://our.internmc.facebook.com/intern/diff/D32175961)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175961).
